osquery/tables/applications/darwin/browser_plugins.cpp

```diff
@@ -227,26 +281,204 @@ QueryData genSafariExtensions(QueryContext& context) {
+    if (boost::algorithm::contains(app_dir, app_to_exclude)) {
+      return true;
+    }
+  }
+
+  return false;
 }
+
+inline bool isUserExtension(const fs::path& app_extension_plist,
+                            const fs::path web_extension_plist,
+                            const SandboxedExtensionData& ext_data) {
+  // Gather user extension metainformation
+  pt::ptree app_extension_ptree;
+  pt::ptree web_extension_ptree;
+
+  // Iterate over the list of user extensions and check if there is a match with
+  // extension identifier
+  getPtreeFromPlist(app_extension_plist, app_extension_ptree);
+  for (const auto& entry : app_extension_ptree) {
+    if (!boost::algorithm::contains(entry.first, ext_data.identifier)) {
+      return true;
+    }
+  }
+
+  getPtreeFromPlist(web_extension_plist, web_extension_ptree);
+  for (const auto& entry : web_extension_ptree) {
+    if (!boost::algorithm::contains(entry.first, ext_data.identifier)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+inline void genSafariSandboxedExtensions(const QueryContext& context,
+                                         QueryData& results) {
+  // Getting user data
+  auto users = usersFromContext(context);
+  for (const auto& row : users) {
+    auto uid = row.at("uid");
+    auto user_dir = row.at("directory");
+
+    // Sanity check
+    if (uid.empty() || user_dir.empty()) {
+      continue;
+    }
+
+    // Skipping non-user queries
+    if (!boost::algorithm::starts_with(user_dir, "/Users")) {
+      continue;
+    }
+
+    // We need to get the sandboxed extension data first
+
+    // Check that an extensions directory exists.
+    if (!pathExists(kSafariAppExtensionsPath).ok()) {
+      return;
+    }
+
+    // Getting app directories
+    std::vector<std::string> app_directories;
+    if (!listDirectoriesInDirectory(
+            kSafariAppExtensionsPath, app_directories, false)) {
+      return;
+    }
+
+    // Traverse app directories to obtain app extension data if present
+    SandboxedExtensionsData sandboxed_ext_data;
+    for (const auto& app_directory : app_directories) {
+      std::vector<std::string> app_internal_dirs;
+      if (!listDirectoriesInDirectory(app_directory, app_internal_dirs, true)) {
+        continue;
+      }
+
+      // Get global sandboxed extensions data
+      for (const auto& app_int_dir : app_internal_dirs) {
+        if (boost::algorithm::ends_with(app_int_dir,
+                                        kSafariSandboxedExtensionsPattern)) {
+          getSandboxedExtensionData(
+              app_directory, app_int_dir + "Info.plist", sandboxed_ext_data);
+        }
+      }
+    }
+
+    // Return if no sandboxed extensions were found - No extensions installed
+    if (sandboxed_ext_data.empty()) {
+      return;
+    }
+
+    // We now have the sandboxed extension metainformation, we now need to check
+    // if extensions were installed by current user
```

Comment:
Hrm. This logic seems weird.

If I understand this, it's traversing all of `/Applications`, looking for possible safari extensions. Then it's enumerating the installed extensions for the user's in context, and finally adding the row if present. 

But, I suspect that enumerating /Applications is a lot slower than enumerating the users directory. Does it make sense to reverse them? Then if there are no extensions installed for the users, you can skip the `/Applications` traverse. Though. this is probably rare, I bet most people have some extensions.

Though... Is it interesting to collect the mismatches? How do we list extensions-not-installed? Or extensions configured, but without a matching bundle id? (Or maybe we don't care about those)

Reply:
Thanks for the input. 

I've moved `/Applications` traversing outside the user traversing loop to perform this once. I misunderstood how this works on MacOS